### ISSUE TITLE
Decryption failure on 16380-byte messages

### ISSUE THREAD

**Issue by krizhanovsky** [tempesta-tech/tempesta#1283]
Use following patch for `test_tls_integrity.py` (introduced in https://github.com/tempesta-tech/tempesta-test/pull/103/commits/519f4d8144e223c3b64927f232d26a0e7b4b8604) just to run single test (all previous tests pass):
```diff
--- a/tls/test_tls_integrity.py
+++ b/tls/test_tls_integrity.py
@@ -148,12 +148,12 @@ class Cache(TlsIntegrityTester):
 
     def test_various_req_resp_sizes(self):
         self.start_all()
-        self.common_check(1, 1)
-        self.common_check(19, 19)
-        self.common_check(567, 567)
-        self.common_check(1755, 1755)
-        self.common_check(4096, 4096)
+        #self.common_check(1, 1)
+        #self.common_check(19, 19)
+        #self.common_check(567, 567)
+        #self.common_check(1755, 1755)
+        #self.common_check(4096, 4096)
         self.common_check(16380, 16380)
-        self.common_check(65536, 65536)
-        self.common_check(1000000, 1000000)
+        #self.common_check(65536, 65536)
+        #self.common_check(1000000, 1000000)
```

The test fails with
```
======================================================================
FAIL: test_various_req_resp_sizes (tls.test_tls_integrity.Cache)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/tempesta-test/tls/test_tls_integrity.py", line 156, in test_various_req_resp_sizes
    self.common_check(16380, 16380)
  File "/root/tempesta-test/tls/test_tls_integrity.py", line 66, in common_check
    " (len=%d)" % (req_len, resp_len))
AssertionError: Cannot process request (len=16380) or response (len=16380)

----------------------------------------------------------------------
Ran 1 test in 6.333s

FAILED (failures=1)
```

I built Tempesta FW with `DEBUG=2 DBG_TLS=1 DBG_HTTP_PARSER=1` (enough verbosity for the error message and not to much to hit any timeouts). The log shows following:
```
[18598.582825] [tempesta tls]   handshake message: msglen=233 type=1 hslen=229 read=9
[18598.584577] [tempesta tls]   server state: 0
[18598.585747] [tempesta tls]   found supported point formats extension
[18598.587263] [tempesta tls]   ClientHello: point format selected: 0
[18598.588716] [tempesta tls]   found supported elliptic curves extension
[18598.590551] [tempesta tls]   found session ticket extension
[18598.591996] [tempesta tls]   found signature_algorithms extension
[18598.593539] [tempesta tls]   ClientHello: signature_algorithm ext: hash alg 5 not supported
[18598.593540] [tempesta tls]   ClientHello: signature_algorithm ext: unknown sig alg encoding 2
[18598.596681] [tempesta tls]   ClientHello: signature_algorithm ext: hash alg 5 not supported
[18598.596682] [tempesta tls]   ClientHello: signature_algorithm ext: match sig 1 and hash 4
[18598.598601] [tempesta tls]   ClientHello: signature_algorithm ext: unknown sig alg encoding 2
[18598.602939] [tempesta tls]   ClientHello: signature_algorithm ext: match sig 4 and hash 4
[18598.602941] [tempesta tls]   ClientHello: signature_algorithm ext: match sig 1 and hash 3
[18598.605076] [tempesta tls]   ClientHello: signature_algorithm ext: unknown sig alg encoding 2
[18598.608987] [tempesta tls]   ClientHello: signature_algorithm ext: match sig 4 and hash 3
[18598.608988] [tempesta tls]   ClientHello: signature_algorithm ext: hash alg 2 not supported
[18598.610727] [tempesta tls]   ClientHello: signature_algorithm ext: unknown sig alg encoding 2
[18598.614195] [tempesta tls]   ClientHello: signature_algorithm ext: hash alg 2 not supported
[18598.614196] [tempesta tls]   ClientHello: signature_algorithm ext: unknown hash alg encoding 2
[18598.617592] [tempesta tls]   ClientHello: signature_algorithm ext: unknown sig alg encoding 2
[18598.619406] [tempesta tls]   ClientHello: signature_algorithm ext: unknown hash alg encoding 2
[18598.620840] [tempesta tls]   unknown extension found: 22 (ignoring)
[18598.621986] [tempesta tls]   found extended master secret extension
[18598.623234] [tempesta tls]   trying ciphersuite: TLS-ECDHE-ECDSA-WITH-AES-128-GCM-SHA256
[18598.624606] [tempesta tls]   ciphersuite requires certificate
[18598.625654] [tempesta tls]   selected ciphersuite: TLS-ECDHE-ECDSA-WITH-AES-128-GCM-SHA256
[18598.627114] [tempesta tls]     update checksum on buf ffff8935dc7fc2e5 len=233, hash=3
[18598.628475] [tempesta tls]   server hello, chosen version 3:3, buf=ffff89358f93c600
[18598.629751] [tempesta tls]   ServerHello: session id len 32
[18598.631302] [tempesta tls]   ServerHello: no session has been resumed
[18598.633432] [tempesta tls]   ServerHello: chosen ciphersuite: TLS-ECDHE-ECDSA-WITH-AES-128-GCM-SHA256
[18598.635295] [tempesta tls]   ServerHello: adding extended master secret extension
[18598.636758] [tempesta tls]   ServerHello: supported_point_formats extension
[18598.638050] [tempesta tls]   ServerHello: total extension length: 10
[18598.640008] [tempesta tls]   write record: type=22 len=86 hslen=0 sgt=ffff8935ffd03668/1 sg_i=0
[18598.642539] [tempesta tls]     update checksum on buf ffff89358f93c600 len=86, hash=3
[18598.644951] [tempesta tls]   write record: type=22 len=695 hslen=0 sgt=ffff8935ffd03668/3 sg_i=1
[18598.647096] [tempesta tls]     update checksum on buf ffff89358f93c65b len=10, hash=3
[18598.648994] [tempesta tls]     update checksum on buf ffff89358f778000 len=685, hash=3
[18598.650843] [tempesta tls]   ECDHE curve: secp256r1
[18598.654691] [tempesta tls]   pick hash algorithm 4 for signing
[18598.658737] [tempesta tls]   write record: type=22 len=148 hslen=0 sgt=ffff8935ffd03668/4 sg_i=3
[18598.660450] [tempesta tls]     update checksum on buf ffff89358f93c66a len=148, hash=3
[18598.663297] [tempesta tls]   sending ServerHelloDone
[18598.664920] [tempesta tls]   write record: type=22 len=4 hslen=0 sgt=ffff8935ffd03668/5 sg_i=4
[18598.666913] [tempesta tls]     update checksum on buf ffff89358f93c703 len=4, hash=3
[18598.668657] [tempesta fw]   TLS 5 bytes +5 segments (4 bytes, last msgtype 0x16) are to be sent on conn=ffff8935c223d040/sk_write_xmit=ffffffffc076ca60 ready=0
[18598.672666] [tempesta tls]   handshake message: msglen=70 type=16 hslen=66 read=9
[18598.674624] [tempesta tls]   server state: 7000000
[18598.675974] [tempesta tls]     update checksum on buf ffff89358f93c2e5 len=70, hash=3
[18598.679954] [tempesta tls]   ciphersuite = TLS-ECDHE-ECDSA-WITH-AES-128-GCM-SHA256
[18598.681363] [tempesta tls]   keylen=16 minlen=24 ivlen=12 maclen=0 tagsize=16 mac_key_len=0
[18598.683228] [tempesta tls]   skip parse certificate verify
[18598.684413] [tempesta tls]   change cipher spec message: msglen=1 type=1 hslen=0 read=81
[18598.685804] [tempesta tls]   server state: 9000000
[18598.686948] [tempesta tls]   server state: a000000
[18598.688050] [tempesta tls]   ttls_parse_finished: msglen=40(rlen=0 len=32)
[18598.689318] [tempesta tls]     decrypt input record from network: hdr=ffff8935c223d3a0 msglen=40 chunks=0 taglen=16 eiv_len=8
[18598.692477] [tempesta tls]     update checksum on buf ffff8935c4693b70 len=16, hash=3
[18598.694263] [tempesta tls]   write record: type=20 len=1 hslen=1 sgt=0000000000000000/0 sg_i=0
[18598.696034] [tempesta fw]   TLS 6 bytes +1 segments (40 bytes, last msgtype 0x16) are to be sent on conn=ffff8935c223d040/sk_write_xmit=ffffffffc076ca60 ready=1
[18598.702621] [tempesta tls]     decrypt input record from network: hdr=ffff8935c223d3a0 msglen=16408 chunks=2 taglen=16 eiv_len=8
[18598.706290] [tempesta tls]   bad message length 16408
[18598.707880] [tempesta tls]   cannot decrypt msg on state e000000, ret=-2
[18598.709709] [tempesta fw]     tfw_http_conn_drop: conn=[ffff8935c223d040]
[18598.712375] [tempesta fw]     tfw_http_conn_cli_drop: conn=[ffff8935c223d040]
[18598.714402] [tempesta fw]     put client ffff8935eb405018, users=8
[18598.716640] [tempesta fw]     put client ffff8935eb405018, users=7
[18603.587584] [tempesta fw]     tfw_http_conn_drop: conn=[ffff8935c223d788]
[18603.589187] [tempesta fw]     tfw_http_conn_cli_drop: conn=[ffff8935c223d788]
[18603.590659] [tempesta fw]     put client ffff8935eb405018, users=6
[18603.592223] [tempesta fw]     put client ffff8935eb405018, users=5
[18603.594858] [tempesta fw]     tfw_http_conn_drop: conn=[ffff8935af2ce080]
[18603.596305] [tempesta fw]     tfw_http_conn_cli_drop: conn=[ffff8935af2ce080]
[18603.597978] [tempesta fw]     put client ffff8935eb405018, users=4
[18603.599754] [tempesta fw]     put client ffff8935eb405018, users=3
[18603.602382] [tempesta fw]     tfw_http_conn_drop: conn=[ffff8935af2ce7c8]
[18603.603787] [tempesta fw]     tfw_http_conn_cli_drop: conn=[ffff8935af2ce7c8]
[18603.605241] [tempesta fw]     put client ffff8935eb405018, users=2
[18603.606716] [tempesta fw]     put client ffff8935eb405018, users=1
[18603.608033] [tempesta fw]   put client: cli=ffff8935eb405018
[18603.659463] [tempesta fw]     got state via sysctl: stop
[18603.660818] [tempesta fw] Stopping all modules...
```

**Comment by i-rinat** [comment]
> bad message length 16408

Reproduced the bug on my machine.

Turns out, it's not an issue with Tempesta, but with the test. Somehow the test actually sends a TLS record with 16408 in `length` field, while [RFC5246](https://tools.ietf.org/html/rfc5246) explicitly states that there is a hard limit of 2¹⁴ (16384) bytes. So we need somehow persuade test scripts to obey TLS 1.2 spec.

There is another issue with the test scripts. As Tempesta likes to emit large TLS records that span multiple packets, it's possible to get `ssl.SSLWantReadError` exceptions in reading code in Python from time to time. These are not actually errors and may be safely ignored.

**Comment by krizhanovsky** [comment]
Need to explore upper limit for [protection records](https://tools.ietf.org/html/rfc5246#section-6.2.3) (any AEAD?) and possible extensions [RFC 8449](https://tools.ietf.org/html/rfc8449) and [RFC 6066 4](https://tools.ietf.org/html/rfc6066#section-4).

**Comment by i-rinat** [comment]
Yeah, looks like I confused plaintext length limitations with ciphertext length limitations. They are a bit different.

**Comment by i-rinat** [comment]
#1322 makes changes to limit checking, so that 16408-byte ciphertexts are now considered valid. THe test with 16380-byte body now passes. However the test with 65536-byte body fails for a non-related reason — there are issues with encryption of responses from the cache.

**Comment by i-rinat** [comment]
#1328 fixes encryption of larger response bodies, thus fixing the other part of this issue.

**Comment by ikoveshnikov** [comment]
Just caught on CI:
```sh
Start test: tls.test_tls_integrity.Cache.test_various_req_resp_sizes
[  481.259206] [tdb] Start Tempesta DB
[  481.275017] [tempesta fw] Initializing Tempesta FW kernel module...
[  481.287871] [tempesta fw] Warning: cache_methods: non-cacheable method 'POST' is set as cacheable
[  481.291604] [tempesta fw] Configuration processing is completed.
[  481.322746] [tdb] Opened table /opt/tempesta/db/filter.tdb: size=16777216 rec_size=20 base=ffff96916ac00000
[  481.466482] [tdb] Opened table /opt/tempesta/db/cache.tdb: size=268435456 rec_size=0 base=ffff96915ac00000
[  481.477418] [tdb] Opened table /opt/tempesta/db/client.tdb: size=16777216 rec_size=616 base=ffff969159c00000
[  481.481562] [tempesta fw] Open listen socket on: 0.0.0.0:443
[  481.500048] [tempesta fw] modules are started
[  481.988581] ------------[ cut here ]------------
[  481.991090] kernel BUG at /root/tempesta/tempesta/tempesta_fw/http.c:3058!
[  481.994203] invalid opcode: 0000 [#1] SMP PTI
[  481.996470] Modules linked in: tempesta_fw(O) tempesta_db(O) tempesta_tls(O) tempesta_lib(O) sha256_ssse3 sha512_ssse3 sha512_generic ccm fuse ata_generic intel_rapl sb_edac crct10dif_pclmul crc32_pclmul ghash_clmulni_intel ata_piix cirrus libata ttm joydev xen_netfront drm_kms_helper scsi_mod intel_rapl_perf drm i2c_piix4 psmouse pcspkr floppy button ip_tables x_tables autofs4 ext4 crc16 mbcache jbd2 crc32c_generic fscrypto ecb crc32c_intel aesni_intel xen_blkfront aes_x86_64 crypto_simd cryptd glue_helper evdev serio_raw [last unloaded: tempesta_lib]
[  482.017710] CPU: 0 PID: 0 Comm: swapper/0 Tainted: G           O    4.14.0-tempesta-amd64 #1 Debian 4.14.32-tfw6-1
[  482.021995] Hardware name: Xen HVM domU, BIOS 4.8.5 01/11/2019
[  482.024898] task: ffffffffa7412480 task.stack: ffffffffa7400000
[  482.027815] RIP: 0010:tfw_http_msg_process_generic+0x968/0xd00 [tempesta_fw]
[  482.031096] RSP: 0018:ffff96918f003970 EFLAGS: 00010216
[  482.034019] RAX: 0000000000000007 RBX: ffff969184148750 RCX: 0000000000000010
[  482.037628] RDX: ffff9690b7778500 RSI: ffff96918f003b50 RDI: ffff969183bae100
[  482.040942] RBP: ffff969185383020 R08: 0000000000000001 R09: 0000000000000000
[  482.044604] R10: 0000000000000000 R11: ffff96912de8a140 R12: ffff96918f003b50
[  482.047900] R13: 0000000000000000 R14: 0000000000000000 R15: ffff969183bae100
[  482.051191] FS:  0000000000000000(0000) GS:ffff96918f000000(0000) knlGS:0000000000000000
[  482.054782] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[  482.057643] CR2: 00007fd47a055360 CR3: 000000010160a005 CR4: 00000000001606f0
[  482.060952] Call Trace:
[  482.062778]  <IRQ>
[  482.064465]  ? irq_exit+0x3c/0xb0
[  482.066540]  ? xen_evtchn_do_upcall+0x2c/0x40
[  482.068951]  tfw_http_msg_process+0x9e/0xe0 [tempesta_fw]
[  482.071675]  ? pg_skb_alloc+0x253/0x470
[  482.073908]  __gfsm_fsm_exec+0x56/0x90 [tempesta_fw]
[  482.076479]  ? skb_split+0x1ff/0x2e0
[  482.078620]  tfw_gfsm_move+0x132/0x160 [tempesta_fw]
[  482.081186]  tfw_tls_msg_process+0x1de/0x360 [tempesta_fw]
[  482.083912]  __gfsm_fsm_exec+0x56/0x90 [tempesta_fw]
[  482.086478]  tfw_connection_recv+0x4e/0x70 [tempesta_fw]
[  482.089167]  ? tfw_connection_send+0x30/0x30 [tempesta_fw]
[  482.091905]  ss_tcp_process_data+0x1db/0x440 [tempesta_fw]
[  482.094653]  ss_tcp_data_ready+0x43/0x90 [tempesta_fw]
[  482.097300]  tcp_rcv_established+0x4d2/0x570
[  482.099646]  tcp_v4_do_rcv+0x129/0x1d0
[  482.101812]  tcp_v4_rcv+0x947/0xa50
[  482.103884]  ip_local_deliver_finish+0x9a/0x1c0
[  482.106272]  ip_local_deliver+0x6b/0xe0
[  482.108427]  ? tcp_v4_early_demux+0x112/0x150
[  482.110740]  ? ip_rcv_finish+0x17a/0x400
[  482.112899]  ip_rcv+0x289/0x3c0
[  482.114806]  ? inet_del_offload+0x40/0x40
[  482.116986]  __netif_receive_skb_core+0x84f/0xb30
[  482.119382]  ? process_backlog+0xa3/0x160
[  482.121629]  process_backlog+0xa3/0x160
[  482.123881]  net_rx_action+0x28e/0x3f0
[  482.126113]  ? handle_irq_event+0x47/0x60
[  482.128409]  __do_softirq+0x10f/0x2a8
[  482.130549]  irq_exit+0xae/0xb0
[  482.132460]  xen_evtchn_do_upcall+0x2c/0x40
[  482.134739]  xen_hvm_callback_vector+0x7d/0x90
[  482.137112]  </IRQ>
[  482.138595] RIP: 0010:native_safe_halt+0x2/0x10
[  482.140908] RSP: 0018:ffffffffa7403e98 EFLAGS: 00000246 ORIG_RAX: ffffffffffffff0c
[  482.144313] RAX: ffffffffa6a97e60 RBX: ffffffffa7412480 RCX: 0000000000000000
[  482.147557] RDX: 0000000000000000 RSI: 0000000000000000 RDI: 0000000000000000
[  482.150799] RBP: 0000000000000000 R08: 000000002062969d R09: ffff9691840c6b00
[  482.153977] R10: 0000000000000000 R11: 0000013d71d7a7fb R12: ffffffffa7412480
[  482.156861] R13: ffffffffa7412480 R14: 0000000000000000 R15: 0000000000000000
[  482.159741]  ? __sched_text_end+0x3/0x3
[  482.161619]  default_idle+0x1a/0xf0
[  482.163367]  do_idle+0x16e/0x1f0
[  482.165034]  cpu_startup_entry+0x6f/0x80
[  482.166897]  start_kernel+0x462/0x482
[  482.168673]  secondary_startup_64+0xa5/0xb0
[  482.170595] Code: 00 00 49 8b 95 80 00 00 00 48 8d b5 c8 00 00 00 e8 6e 86 fe ff 84 c0 0f 84 d4 fd ff ff f0 41 80 a5 a9 00 00 00 fd e9 c6 fd ff ff <0f> 0b 83 f8 03 75 32 41 b8 01 00 00 00 b9 01 00 00 00 48 c7 c2 
[  482.177688] RIP: tfw_http_msg_process_generic+0x968/0xd00 [tempesta_fw] RSP: ffff96918f003970
[  482.181331] ---[ end trace 8b7440df3e3c1a78 ]---
[  482.183485] Kernel panic - not syncing: Fatal exception in interrupt
[  482.186209] Kernel Offset: 0x25400000 from 0xffffffff81000000 (relocation range: 0xffffffff80000000-0xffffffffbfffffff)
program finished with exit code 0
```

But I couldn't reproduce it on my VM locally.

**Comment by i-rinat** [comment]
> kernel BUG at /root/tempesta/tempesta/tempesta_fw/http.c:3058

Fixed by #1330.
